### ISSUE TITLE
Contribution Guidelines

### ISSUE THREAD

**Issue by bocajnotnef** [mercere99/Empirical#6]
I think this was mentioned in a verbal conversation though I don't remember quite which.

I feel one of the early pieces of documentation should be contribution guidelines for the project, detailing preferred formatting specs, if/how commits should be squashed, how PRs should be reviewed, etc. I think that establishing that early on will prevent problems down the line.

Additionally if we establish clear rules for what's required in a PR/merge then it's easier to prevent the review process from blocking on one specific person, or at least reducing the amount of time required by reviewers with a high value-to-time ration (i.e. @mercere99)

Some problems with this:
 * We currently lack automated formatting/testing, making it difficult to easily keep to a standard ("write code to enforce policy" and all that)
 * We don't currently have a decided formatting standard anyway and I'm not sure how important an issue that should be

Thoughts/Comments?

**Comment by bocajnotnef** [comment]
Shamelessly plugging the khmer contributor checklist--It's decently thorough and can be easily adapted to our needs.

https://khmer.readthedocs.org/en/v1.4.1/dev/coding-guidelines-and-review.html#checklist

**Comment by mercere99** [comment]
I just wrote this elsewhere, but here is more relevant; we do have some coding guidelines written up, though a lot more are needed: https://github.com/mercere99/Empirical/wiki/Coding-Guidelines

My thought is that we want these pieces in place before we open up the project further for more people to get involved, but it's okay if we take the time to put the testing framework together a bit more first.

I do think that checklists will be important, and the khmer ones are a good starting point.

**Comment by bocajnotnef** [comment]
+1 for getting testing in place before opening up more. I think requiring
teat coverage for code addition/making sure tests pass for contributions is
a good thing to have.

**Comment by bocajnotnef** [comment]
Additional requirement: copyright header/attribution to the Empirical project

**Comment by bocajnotnef** [comment]
Also, ref #3

**Comment by bocajnotnef** [comment]
Cool. So, I copied the wiki page for Empirical formatting into docs within Empirical itself and added a (short) review checklist (https://github.com/mercere99/Empirical/commit/55e98f1ac8d8f8dc9d80999c5f8c031ca0c54e9f)

Open to thoughts/comments/suggestions

Going to work on finalizing the Catch implementation and then I'll write some maketargets to do testing/format checking/etc.

What'll be interesting is writing dependency installation notes for all the things I've added (catch, sphinx, breathe, doxygen).
